File: app/src/main/java/com/soho/sohoapp/feature/home/editproperty/data/PropertyListing.java
package com.soho.sohoapp.feature.home.editproperty.data;

import android.os.Parcel;
import android.os.Parcelable;

public class PropertyListing implements Parcelable {
    private int id;
    private String state;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.id);
        dest.writeString(this.state);
    }

    public PropertyListing() {
        //empty constructor is required because you will not be able to create object
    }

    protected PropertyListing(Parcel in) {
        this.id = in.readInt();
        this.state = in.readString();
    }

    public static final Creator<PropertyListing> CREATOR = new Creator<PropertyListing>() {
        @Override
        public PropertyListing createFromParcel(Parcel source) {
            return new PropertyListing(source);
        }

        @Override
        public PropertyListing[] newArray(int size) {
            return new PropertyListing[size];
        }
    };
}

File: app/src/main/java/com/soho/sohoapp/data/models/PropertyListing.kt
package com.soho.sohoapp.data.models

import android.os.Parcel
import android.os.Parcelable
import com.soho.sohoapp.data.enums.PropertyStatus

class PropertyListing() : Parcelable {

    var id: Int = 0
    var state: String? = null
    var canReceiveSalesOffers: Boolean = false
    var canReceiveRentOffers: Boolean = false
    var saleTitle: String? = null
    var rentTitle: String? = null
    var auctionTitle: String? = null
    var discoverableTitle: String? = null
    var isOnSiteAuction: Boolean = false
    var auctionTime: Long? = null
    var rentPaymentFrequency: String? = null
    var availableFrom: Long? = null
    var isAppointmentOnly: Boolean = false
    var offSiteLocation: Location? = null
    var inspectionTimes: List<InspectionTime>? = null

    val isPublic: Boolean
        get() = PropertyStatus.DISCOVERABLE == state ||
                PropertyStatus.SALE == state ||
                PropertyStatus.AUCTION == state ||
                PropertyStatus.RENT == state

    val isPrivate: Boolean
        get() = PropertyStatus.PRIVATE == state

    val isArchived: Boolean
        get() = PropertyStatus.ARCHIVED == state || PropertyStatus.SOLD == state

    constructor(parcel: Parcel) : this() {
        id = parcel.readInt()
        state = parcel.readString()
        canReceiveSalesOffers = parcel.readByte() != 0.toByte()
        canReceiveRentOffers = parcel.readByte() != 0.toByte()
        saleTitle = parcel.readString()
        rentTitle = parcel.readString()
        auctionTitle = parcel.readString()
        discoverableTitle = parcel.readString()
        isOnSiteAuction = parcel.readByte() != 0.toByte()
        auctionTime = parcel.readValue(Long::class.java.classLoader) as Long?
        rentPaymentFrequency = parcel.readString()
        availableFrom = parcel.readValue(Long::class.java.classLoader) as Long?
        isAppointmentOnly = parcel.readByte() != 0.toByte()
        offSiteLocation = parcel.readParcelable(Location::class.java.classLoader)
        inspectionTimes = parcel.createTypedArrayList(InspectionTime)
    }

    fun getInspectionTimesSafe() = inspectionTimes ?: ArrayList<InspectionTime>()

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(id)
        parcel.writeString(state)
        parcel.writeByte(if (canReceiveSalesOffers) 1 else 0)
        parcel.writeByte(if (canReceiveRentOffers) 1 else 0)
        parcel.writeString(saleTitle)
        parcel.writeString(rentTitle)
        parcel.writeString(auctionTitle)
        parcel.writeString(discoverableTitle)
        parcel.writeByte(if (isOnSiteAuction) 1 else 0)
        parcel.writeValue(auctionTime)
        parcel.writeString(rentPaymentFrequency)
        parcel.writeValue(availableFrom)
        parcel.writeByte(if (isAppointmentOnly) 1 else 0)
        parcel.writeParcelable(offSiteLocation, flags)
        parcel.writeTypedList(inspectionTimes)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<PropertyListing> {
        override fun createFromParcel(parcel: Parcel): PropertyListing {
            return PropertyListing(parcel)
        }

        override fun newArray(size: Int): Array<PropertyListing?> {
            return arrayOfNulls(size)
        }
    }


}
